homeassistant/components/logi_circle/__init__.py

```diff
@@ -2,74 +2,220 @@
+        client_secret=entry.data[CONF_CLIENT_SECRET],
+        api_key=entry.data[CONF_API_KEY],
+        redirect_uri=entry.data[CONF_REDIRECT_URI],
+        cache_file=DEFAULT_CACHEDB
+    )
+
+    if not logi_circle.authorized:
         hass.components.persistent_notification.create(
-            'Error: {}<br />'
-            'You will need to restart hass after fixing.'
-            ''.format(ex),
+            "Error: The cached access tokens are missing from {}.<br />"
+            "Please unload then re-add the Logi Circle integration to resolve."
+            ''.format(DEFAULT_CACHEDB),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    try:
+        with async_timeout.timeout(_TIMEOUT, loop=hass.loop):
+            # Ensure the cameras property returns the same Camera objects for
+            # all devices. Performs implicit login and session validation.
+            await logi_circle.synchronize_cameras()
+    except AuthorizationFailed:
+        hass.components.persistent_notification.create(
+            "Error: Failed to obtain an access token from the cached "
+            "refresh token.<br />"
+            "Token may have expired or been revoked.<br />"
+            "Please unload then re-add the Logi Circle integration to resolve",
             title=NOTIFICATION_TITLE,
             notification_id=NOTIFICATION_ID)
         return False
     except asyncio.TimeoutError:
         # The TimeoutError exception object returns nothing when casted to a
         # string, so we'll handle it separately.
-        err = '{}s timeout exceeded when connecting to Logi Circle API'.format(
+        err = "{}s timeout exceeded when connecting to Logi Circle API".format(
             _TIMEOUT)
-        _LOGGER.error(err)
         hass.components.persistent_notification.create(
-            'Error: {}<br />'
-            'You will need to restart hass after fixing.'
+            "Error: {}<br />"
+            "You will need to restart hass after fixing."
             ''.format(err),
             title=NOTIFICATION_TITLE,
             notification_id=NOTIFICATION_ID)
         return False
+    except ClientResponseError as ex:
+        hass.components.persistent_notification.create(
+            "Error: {}<br />"
+            "You will need to restart hass after fixing."
+            ''.format(ex),
+            title=NOTIFICATION_TITLE,
+            notification_id=NOTIFICATION_ID)
+        return False
+
+    hass.data[DATA_LOGI] = {
+        ATTR_API: logi_circle,
+        'entities': {
+            'camera': []
+        }
+    }
+
+    for component in 'camera', 'sensor':
+        hass.async_create_task(hass.config_entries.async_forward_entry_setup(
+            entry, component))
+
+    async def service_handler(service):
+        """Dispatch service calls to target entities."""
+        cameras = hass.data[DATA_LOGI]['entities']['camera']
```

Comment:
We're not allowed to access entities outside their platforms. Use our dispatch helper to signal the correct entity.

Reply:
Addressed in commit [3aaedb86](https://github.com/home-assistant/home-assistant/pull/20624/commits/3aaedb86d735509f2ab939668741aeb0ce10b973).